File: py/src/lib/stdlib.py
"""
Standard Library Functions
"""

import random
import time
import sys

class Stdlib:
    @staticmethod
    def malloc(size):
        """Allocate memory"""
        print(f"[STDLIB] malloc({size})")
        # In real implementation, this would interface with memory manager
        return [0] * size
    
    @staticmethod
    def free(ptr):
        """Free memory"""
        print(f"[STDLIB] free({id(ptr)})")
        # In real implementation, this would return memory to memory manager
    
    @staticmethod
    def calloc(nmemb, size):
        """Allocate and zero memory"""
        total_size = nmemb * size
        print(f"[STDLIB] calloc({nmemb}, {size}) = {total_size}")
        return [0] * total_size
    
    @staticmethod
    def realloc(ptr, size):
        """Reallocate memory"""
        print(f"[STDLIB] realloc({id(ptr)}, {size})")
        # Simplified implementation
        return [0] * size
    
    @staticmethod
    def atoi(s):
        """String to integer"""
        try:
            return int(s)
        except ValueError:
            return 0
    
    @staticmethod
    def atol(s):
        """String to long"""
        try:
            return int(s)
        except ValueError:
            return 0
    
    @staticmethod
    def rand():
        """Random number"""
        return random.randint(0, 32767)
    
    @staticmethod
    def srand(seed):
        """Seed random number generator"""
        random.seed(seed)
    
    @staticmethod
    def exit(status):
        """Exit program"""
        print(f"[STDLIB] exit({status})")
        sys.exit(status)
    
    @staticmethod
    def system(command):
        """Execute system command"""
        print(f"[STDLIB] system('{command}')")
        # In real OS, this would create a new process
        return 0
    
    @staticmethod
    def getenv(name):
        """Get environment variable"""
        # Simulated environment
        env_vars = {
            'PATH': '/bin:/usr/bin',
            'HOME': '/home/user',
            'USER': 'user',
            'SHELL': '/bin/sh'
        }
        return env_vars.get(name, "")
    
    @staticmethod
    def abs(x):
        """Absolute value"""
        return abs(x)
    
    @staticmethod
    def labs(x):
        """Long absolute value"""
        return abs(x)

File: c/src/lib/stdlib.c
#include "lib/stdlib.h"

static unsigned long next = 1;

int rand(void) {
    next = next * 1103515245 + 12345;
    return (unsigned int)(next / 65536) % 32768;
}

void srand(unsigned int seed) {
    next = seed;
}

void* malloc(size_t size) {
    return kmalloc(size);
}

void free(void* ptr) {
    kfree(ptr);
}

void* calloc(size_t num, size_t size) {
    void* ptr = kmalloc(num * size);
    if(ptr) {
        kmemset(ptr, 0, num * size);
    }
    return ptr;
}

void* realloc(void* ptr, size_t new_size) {
    if(!ptr) return kmalloc(new_size);
    
    // Simple implementation - always allocate new block
    void* new_ptr = kmalloc(new_size);
    if(new_ptr) {
        // Copy old data (would need to know old size in real implementation)
        kmemcpy(new_ptr, ptr, new_size); // This is unsafe without old size
        kfree(ptr);
    }
    return new_ptr;
}

int abs(int x) {
    return x < 0 ? -x : x;
}

long labs(long x) {
    return x < 0 ? -x : x;
}

int atexit(void (*func)(void)) {
    // Simple implementation - not supported
    return 0;
}

void exit(int status) {
    // Would call task_exit in real implementation
    while(1) asm volatile("hlt");
}